src/v/coproc/wasm_event_listener.cc

```diff
@@ -0,0 +1,216 @@
+ss::future<>
+wasm_event_listener::resolve_wasm_script(ss::sstring name, iobuf source_code) {
+    std::filesystem::path active_path(_active_dir / name.c_str());
+    return ss::file_exists(active_path.string())
+      .then([this,
+             active_path,
+             name = std::move(name),
+             source_code = std::move(source_code)](bool exists) mutable {
+          if (query_action(source_code) == wasm_event_action::remove) {
+              if (!exists) {
+                  vlog(
+                    coproclog.warn,
+                    "Attempt to remove coprocessor {} but it doesn't exist in "
+                    "the active directory {}",
+                    name,
+                    active_path);
+                  return ss::now();
+              }
+              std::filesystem::path mvto(_inactive_dir / name.c_str());
+              vlog(
+                coproclog.info,
+                "Moving coprocessor {} to inactive dir",
+                active_path);
+              return ss::rename_file(active_path.string(), mvto.string());
+          } else {
+              if (exists) {
+                  vlog(
+                    coproclog.warn,
+                    "Attempt to deploy coprocessor with script {} but one "
+                    "already exists: {}",
+                    name,
+                    active_path);
+                  return ss::now();
+              }
+              vlog(
+                coproclog.info,
+                "Writing coprocessor {} to submit dir {}",
+                name,
+                _submit_dir);
+              std::filesystem::path submit_src(_submit_dir / name.c_str());
+              return write_fully(submit_src, std::move(source_code));
+          }
+      });
+}
+
+wasm_event_listener::wasm_event_listener(std::filesystem::path wasm_root)
+  : _client(
+    std::vector<unresolved_address>{unresolved_address("127.0.0.1", 9092)})
+  , _wasm_root(std::move(wasm_root))
+  , _active_dir(_wasm_root / "active")
+  , _submit_dir(_wasm_root / "submit")
+  , _inactive_dir(_wasm_root / "inactive") {}
+
+void wasm_event_listener::start() {
+    if (_gate.is_closed() || _gate.get_count() != 0) {
+        throw std::logic_error(
+          "Attempted to start() the wasm_event_notifier run "
+          "loop, after it has already been started or closed");
+    }
+    (void)ss::with_gate(_gate, [this] {
+        return storage::directories::initialize(_active_dir.string())
+          .then([this] {
+              return storage::directories::initialize(_inactive_dir.string());
+          })
+          .then([this] {
+              return storage::directories::initialize(_submit_dir.string());
+          })
+          .then([this] { return _client.connect(); })
+          .then([this] {
+              return ss::do_until(
+                [this] { return _abort_source.abort_requested(); },
+                [this] { return do_start(); });
```

Comment:
`do_start` in a loop is sort of confusing. is it doing a unit of work?

Reply:
I guess you could say that. There is a ss::repeat within do_start() but it doesn't loop indefinitely. It reads some data , performs work then maybe sleeps. After that the future is completed and the abort source in this do_until will be checked